File: examples/quickstart-typescript/contract/checker_impls/validate-sku.ts
/**
 * Custom checker: validate that all item SKUs match the expected format.
 *
 * Stele calls check(steleContext) when evaluating the SKU_FORMAT invariant.
 * The checker reads steleContext.items — a flat list of item objects wired
 * in tests/contract/stele_context.ts.
 */

/** SKU format: one or more uppercase letters, a dash, then alphanumerics. */
const SKU_PATTERN = /^[A-Z]+-[A-Z0-9]+$/;

export function check(
  steleContext: Record<string, unknown>,
  _kwargs: Record<string, unknown>,
): { passed: boolean; message: string | null } {
  const items = Array.isArray(steleContext["items"])
    ? (steleContext["items"] as Record<string, unknown>[])
    : [];
  for (const item of items) {
    const sku = String(item["sku"] ?? "");
    if (!SKU_PATTERN.test(sku)) {
      return {
        passed: false,
        message: `SKU ${JSON.stringify(sku)} does not match ^[A-Z]+-[A-Z0-9]+$`,
      };
    }
  }
  return { passed: true, message: null };
}

File: examples/quickstart-python/contract/checker_impls/validate_sku.py
"""
Custom checker: validate that all item SKUs match the expected format.

Stele calls check(stele_context) when evaluating the SKU_FORMAT invariant.
The checker reads stele_context["items"] — a flat list of item dicts wired
in tests/contract/conftest.py.
"""

import re

# SKU format: one or more uppercase letters, a dash, then alphanumerics.
# Examples: WIDGET-A1, GADGET-C3, PRODUCT-123
SKU_PATTERN = re.compile(r"^[A-Z]+-[A-Z0-9]+$")


def check(stele_context, **_kwargs):
    """
    Returns {"passed": bool, "message": str | None}.

    Iterates over stele_context["items"] and fails on the first invalid SKU.
    """
    items = stele_context.get("items", [])
    for item in items:
        sku = item.get("sku", "") if isinstance(item, dict) else getattr(item, "sku", "")
        if not SKU_PATTERN.match(str(sku)):
            return {
                "passed": False,
                "message": f"SKU {sku!r} does not match ^[A-Z]+-[A-Z0-9]+$",
            }
    return {"passed": True, "message": None}
